src/v/iceberg/compatibility.cc

```diff
@@ -501,4 +501,140 @@ try_fill_field_ids(const struct_type& source, struct_type& dest) {
     }
 }
 
+namespace {
+
+struct merging_schema_visitor {
+    schema_merge_result
+    operator()(const struct_type& source, struct_type& dest) const {
+        for (const auto& f : source.fields) {
+            auto matches = dest.fields
+                           | std::views::filter(
+                             [&f](const nested_field_ptr& df) {
+                                 return df != nullptr && df->name == f->name;
+                             });
+
+            auto match_it = matches.begin();
+            auto n_matches = std::distance(match_it, matches.end());
+
+            switch (n_matches) {
+            case 0:
+                dest.fields.push_back(f->copy());
+                dest.fields.back()->id = nested_field::id_t{};
+                break;
+            case 1: {
+                if ((*match_it)->required == field_required::yes) {
+                    // If the destination field is required the source field is
+                    // allowed to change nullability. I.e. it is allowed to
+                    // become non-required.
+                    (*match_it)->required = f->required;
+                } else if (f->required == field_required::yes) {
+                    // If source field is required but destination field is not,
+                    // we have a conflict.
+                    return schema_evolution_errc::new_required_field;
+                }
+
+                auto field_merge_res = std::visit(
+                  *this, f->type, (*match_it)->type);
+                if (field_merge_res.has_error()) {
+                    return field_merge_res.error();
+                }
+            } break;
+            default:
+                return schema_evolution_errc::ambiguous;
+            }
+        }
+
+        return outcome::success();
+    }
+
+    schema_merge_result
+    operator()(const primitive_type& source, primitive_type& dest) const {
+        auto res = check_types(dest, source);
+        if (res.has_error()) {
+            // If this failed but reverse promotion succeeds we still have a
+            // valid merge result.
+            auto reverse_res = check_types(source, dest);
```

Comment:
not sure I understand what's happening here. why do we need to check primitive type compatibility at this stage? if we get this far, we know there's a name match in the parent, so I would have assumed we return `success` indiscriminately here.